service/auth/session_cache.go

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2022-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package auth
+
+import (
+	"fmt"
+	"sync"
+	"time"
+)
+
+type CachedSession struct {
+	ClientID       string
+	ExpirationDate time.Time
+}
+
+type SessionCacheConfig struct {
+	ExpirationMinutes int `toml:"expiration_minutes"`
+}
+
+func (c SessionCacheConfig) IsValid() error {
+	if c.ExpirationMinutes <= 0 {
+		return fmt.Errorf("invalid ExpirationMinutes value: should be a positive number")
+	}
+	return nil
+}
+
+type SessionCache struct {
+	cfg        *SessionCacheConfig
+	sessionMap map[string]CachedSession
+
+	mut sync.RWMutex
+}
+
+func NewSessionCache(cfg SessionCacheConfig) (*SessionCache, error) {
+	if err := cfg.IsValid(); err != nil {
+		return nil, err
+	}
+	return &SessionCache{cfg: &cfg, sessionMap: make(map[string]CachedSession)}, nil
+}
+
+func (t *SessionCache) Get(token string) (CachedSession, error) {
+	t.mut.RLock()
+	session, ok := t.sessionMap[token]
+	t.mut.RUnlock()
+	if !ok {
+		return CachedSession{}, fmt.Errorf("token is invalid")
+	}
+	if time.Now().After(session.ExpirationDate) {
+		t.mut.Lock()
+		delete(t.sessionMap, token)
+		t.mut.Unlock()
+		return CachedSession{}, fmt.Errorf("session is expired")
+	}
+	return session, nil
+}
+
+func (t *SessionCache) Put(clientID string) (string, error) {
+	if len(clientID) == 0 {
+		return "", fmt.Errorf("can not cache: invalid client id")
+	}
+
+	token, err := t.newUniqueToken()
+	if err != nil {
+		return "", err
+	}
+
+	// Make sure there is only one bearer token per client.
+	t.Delete(clientID)
+
+	t.mut.Lock()
+	t.sessionMap[token] = CachedSession{ClientID: clientID, ExpirationDate: time.Now().Add(time.Duration(t.cfg.ExpirationMinutes) * time.Minute)}
+	t.mut.Unlock()
+	return token, nil
+}
+
+func (t *SessionCache) Delete(clientID string) {
+	t.mut.Lock()
+	defer t.mut.Unlock()
+	for token, session := range t.sessionMap {
+		if session.ClientID == clientID {
+			delete(t.sessionMap, token)
+			return
+		}
+	}
+}
+
+func (t *SessionCache) newUniqueToken() (string, error) {
+	token, err := newRandomString(MinKeyLen)
+	if err != nil {
+		return "", err
+	}
+	t.mut.RLock()
+	defer t.mut.RUnlock()
+	// If the bearer token already exists, try again
+	if _, ok := t.sessionMap[token]; ok {
+		return t.newUniqueToken()
```

Comment:
This looks like it could cause potential recursive locking which is likely not what we want to do here. I'd probably suggest to avoid the `defer` in this case and simply lock/unlock as needed.